checker/src/main/java/org/checkerframework/checker/calledmethods/CalledMethodsVisitor.java

```diff
@@ -59,12 +65,57 @@ public Void visitMethod(MethodTree tree, Void p) {
         checker.report(tree, new DiagMessage(Diagnostic.Kind.ERROR, "ensuresvarargs.invalid"));
       }
     }
+    for (EnsuredCalledMethodOnException postcond :
+        ((CalledMethodsAnnotatedTypeFactory) atypeFactory).getExceptionalPostconditions(elt)) {
+      checkExceptionalPostcondition(postcond, tree);
+    }
     return super.visitMethod(tree, p);
   }
 
+  /**
+   * Check if the given postcondition is really ensured by the body of the given method.
+   *
+   * @param postcond the postcondition to check
+   * @param tree the method
+   */
+  protected void checkExceptionalPostcondition(
+      EnsuredCalledMethodOnException postcond, MethodTree tree) {
+    CFAbstractStore<?, ?> exitStore = atypeFactory.getExceptionalExitStore(tree);
+    if (exitStore == null) {
+      // If there is no exceptional exitStore, then the method cannot throw exceptions and
+      // there is no need to check anything.
+      return;
+    }
+
+    JavaExpression e;
+    try {
+      e = StringToJavaExpression.atMethodBody(postcond.getExpression(), tree, checker);
+    } catch (JavaExpressionParseUtil.JavaExpressionParseException ex) {
+      checker.report(tree, ex.getDiagMessage());
+      return;
+    }
+
+    AnnotationMirror requiredAnno = atypeFactory.createAccumulatorAnnotation(postcond.getMethod());
+
+    CFAbstractValue<?> value = exitStore.getValue(e);
+    AnnotationMirror inferredAnno = null;
+    if (value != null) {
+      AnnotationMirrorSet annos = value.getAnnotations();
+      inferredAnno = qualHierarchy.findAnnotationInSameHierarchy(annos, requiredAnno);
+    }
+
+    if (!checkContract(e, requiredAnno, inferredAnno, exitStore)) {
+      checker.reportError(
+          tree,
+          "contracts.postcondition",
+          tree.getName(),
+          contractExpressionAndType(postcond.getExpression(), inferredAnno),
+          contractExpressionAndType(postcond.getExpression(), requiredAnno));
+    }
```

Comment:
If we have both `@EnsuresCalledMethods` and `@EnsuresCalledMethodsOnException` annotations on a method, for the same called method and value, will it be clear from the error message we get here that the issue is with the exception annotation, not the normal one?

Reply:
Great catch! I'll see what I can do to improve the error message.